Support external Slurm DBD

hanwen-pcluste opened the pull request:
### Description of changes
* Describe *what* you're changing and *why* you're doing these changes.

### Tests
* Describe the automated and/or manual tests executed to validate the patch.
* Describe the added/modified tests.

### References
* Link to impacted open issues.
* Link to related PRs in other packages (i.e. cookbook, node).
* Link to documentation useful to understand the changes.

### Checklist
- Make sure you are pointing to **the right branch**.
- If you're creating a patch for a branch other than `develop` add the branch name as prefix in the PR title (e\.g\. `[release-3.6]`).
- Check all commits' messages are clear, describing what and why vs how.
- Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

hehe7318 committed:
Add entrypoints for external slurmdbd.

hehe7318 committed:
Add cluster_user attribute, add template slurmdbd.service, add template update_munge_key script, add config_munge_key recipe. Use dbms_password_secret_arn and munge_key_secret_arn instead of digging from config when external slurmdbd. Add TODO list

hehe7318 committed:
Create new slurm_external_slurmdbd.conf. Add logic in slurmdbd.conf.erb to use new conf when configing external_slurmdbd.

hehe7318 committed:
Address comments.

hehe7318 committed:
Address issues in External Slurmdbd cookbook recipes (#2585)

Fix Chef::Exceptions::FileNotFound error.

Add logic of create some necessary directories.

jdeamicis committed:
Create slurm_external_slurmdbd.conf for external slurmdbd

Add new Chef template to generate an include file for the
slurmdbd.conf only in case of external slurmdbd configuration.

Minor reformatting of slurmdbd.conf.erb.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Disable Slurm database bootstrap from external slurmdbd

The external slurmdbd instance cannot use Slurm client commands
since it does not have a "full configuration source" (it's
outside a Slurm cluster).

This must be investigated with SchedMD.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Add support for external slurmdbd in Slurm configuration (#2595)

Use ExternalSlurmdbd to render AccountingStorageHost when defined.

Add unit test case to cover external slurmdbd scenario.

Adapt other unit tests for Slurm configuration rendering when Slurm
Accounting is used.

---

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Increase debug level on external slurmdbd.conf

Remove `NoAddrCache` parameter (this should be useless after
a rebase).

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Use static IP to populate slurmdbd's DbdAddr parameter

Export the IP into the instance dna.json.

Add minor comments to the code.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

NSsirena committed:
Configure CloudWatch agent for External Slurm Dbd

Add a dedicated NodeType `ExternalSlurmDbd` that is required
by the recipe logic to trigger the right behavior.

Signed-off-by: Nicola Sirena <nssirena@amazon.com>

jdeamicis committed:
Configure S3 mountpoint for external slurmdbd (#2633)

Install mountpoint-s3 and mount the S3 bucket of the external
slurmdbd stack on `/opt/slurm/etc`.

The slurmdbd configuration files are already generated from
a template with a `:create_if_missing` action, meaning that
if the files are already present in the bucket they will not
be overwritten.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

NSsirena committed:
Add recipe to disable not required services in ExternalSlurmdbd

ParallelCluster AMI are configured for Desktop use and so have some
active service that is not required and are certainly unused
like { apache2, cups, ...,  wpa_supplicant }

Signed-off-by: Nicola Sirena <nssirena@amazon.com>

jdeamicis committed:
Fetch slurmdbd configuration from S3 if available

Remove previously implemented logic to configur s3 mountpoint
for the configuration of the external slurmdbd

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Make DbdPort configurable in external slurmdbd recipes

Do not configure DbdPort in slurmdbd.conf in case of external slurmdbd.

Move configuration of slurmdbd.service into config_slurm_accounting.rb (this is
also called by the external slurmdbd recipe).

Skip some steps in config_slurm_accounting.rb in kitchen tests.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Extend remote_object resource with owner, group and mode

This allows to have a single action to download an item from S3
and assign it an owner, a group or a permission mode, without the
need to specify a separate `file` action, which could generate an
empty file in case the intended remote file does not exist.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Configure External Slurmdbd Host and Port in cookbook

Adapt External Slurmdbd configuration on the PC cluster side
to the Host and Port subparameter of the ExternalSlurmdbd
schema.

Adapt unit tests accordingly.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Fix failing unit tests and cookstyle linter

Add ExternalSlurmdbd subsection in unit tests for the
generation of the Slurm configuration.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

himani2411 commented on cookbooks/aws-parallelcluster-entrypoints/recipes/external_slurmdbd_config.rb:
```diff
@@ -0,0 +1,26 @@
+# OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
+# limitations under the License.
+
+node.default['cluster']['slurm']['install_dir'] = '/opt/slurm'
```
Why not set these in an Attribute file?

himani2411 replied:
I am looking at the PR sequentially by the commit, so some of the comments might be based on the initial commit which you might have changed in the last/later commit.

```# Mock necessary attributes to reuse the recipes from ParallelCluster.```
The above comment about mocking the attributes adds the burden of us maintaining the values in 2 places.

hanwen-pcluste replied:
I deleted the code. It is not needed

himani2411 commented on cookbooks/aws-parallelcluster-entrypoints/recipes/external_slurmdbd_config.rb:
```diff
@@ -18,9 +18,51 @@
+node.default['cluster']['munge']['group'] = node['cluster']['munge']['user']
+
+if platform?('amazon') && node['platform_version'] == "2"
+  node.default['cluster']['cluster_user'] = 'ec2-user'
```
Maybe we should use the `users_amazon2.rb` file

hanwen-pcluste replied:
I deleted the code. It is not needed

himani2411 commented on cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/templates/slurm_external_slurmdbd.conf:
```diff
@@ -0,0 +1,11 @@
+{##}StorageLoc={{ dbms_database_name }}
+{##}StorageUser={{ dbms_username }}
+{##}StoragePass=dummy
+{##}{% endif %}
```
I dont see the start of the if condition

hanwen-pcluste replied:
Done

himani2411 commented on cookbooks/aws-parallelcluster-slurm/recipes/config/config_slurm_accounting.rb:
```diff
@@ -24,6 +24,23 @@
+  mode '0600'
+  action :create_if_missing
+  variables(
+    dbd_host: "localhost",
```
Nitpick: If we are never overriding the dbd_host and storage_port and are always hardcoded to localhost and 3306 respectively, then they should be hardcoded in template itself.

himani2411 commented on cookbooks/aws-parallelcluster-slurm/recipes/config/config_external_slurmdbd_s3_mountpoint.rb:
```diff
@@ -0,0 +1,50 @@
+  action :create_if_missing
+end
+
+if platform?('ubuntu')
```
TODO: add unit test for these recipies.

hanwen-pcluste replied:
This commit is outdated

hanwen-cluster committed:
Improve update_munge_key.sh

1. Directly referencing node attributes
2. Do not add key to "shared folders" on SlurmDBD instances

Signed-off-by: Hanwen <hanwenli@amazon.com>

hanwen-cluster committed:
Remove unnecessary code in external_slurmdbd_config

Signed-off-by: Hanwen <hanwenli@amazon.com>

hanwen-cluster committed:
Simplify slurm_external_slurmdbd.conf.erb

The node attributes can be used in cookbook template directly.

Signed-off-by: Hanwen <hanwenli@amazon.com>

himani2411 commented on cookbooks/aws-parallelcluster-slurm/recipes/config/external_slurmdbd_disable_unrequired_services.rb:
```diff
@@ -0,0 +1,58 @@
+# This was tested on Ubuntu 20.04, Alinux2 and Rocky8
+# Others OS should be checked and added
+
+serviceList = %w()
```
Changing this to a resource rather than a recipe would be better in the longer run.

With a stop and Disable actions

himani2411 replied:
Though this can be an improvement  so that it doesnt block the current PR

himani2411 commented:
LGTM

himani2411 commented on cookbooks/aws-parallelcluster-slurm/recipes/config/config_slurm_accounting.rb:
```diff
@@ -24,6 +32,15 @@
   action :create_if_missing
 end
 
+template "#{node['cluster']['slurm']['install_dir']}/etc/slurm_external_slurmdbd.conf" do
```
doesnt this require a change in Kitchen tests?

himani2411 replied:
Also can we add more Rspec Tests for the recipies we added for external Surmdbd

hanwen-cluster committed:
Remove duplicate attribute node['is_external_slurmdbd']

Signed-off-by: Hanwen <hanwenli@amazon.com>

hanwen-cluster committed:
[Github CI] Add entrypoints spec test to Github checks

Signed-off-by: Hanwen <hanwenli@amazon.com>

hanwen-cluster committed:
Add spec tests for external slurmdbd config recipe

Signed-off-by: Hanwen <hanwenli@amazon.com>
